File: src/admin/server/controller.rs
use num_bigint::BigUint;
use num_traits::identities::{Zero, One};
use std::collections::HashMap;

use sd_project::utils::models::client::DataClient;
use sd_project::utils::request::Request;
use sd_project::utils::response::Response;

pub struct ClientController {
    data: HashMap<String, DataClient>,
    clients: BigUint
}
    
impl ClientController {
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
            clients: BigUint::zero()
        }
    }

    pub fn create(&mut self, req: &Request) -> Response {
        if req.body == None {
            return Response::Error("Incomplet argumets".to_string());
        }

        let data = req.body.as_ref().unwrap();
        match DataClient::from_string(data.to_string()) {
            Ok(data) => {
                let cid = self.clients.to_string();
                self.data.insert(self.clients.to_string(), data);

                self.clients += BigUint::one();

                return Response::Sucess(format!("Client created with sucess! \n CID: {}", cid).to_string());
            },
            Err(e) => Response::Error(e.to_string())
        }
    }

    pub fn update(&mut self, req: &Request) -> Response {
        if req.cid == None || req.body == None {
            return Response::Error("Incomplet argumets".to_string());
        }

        let cid = req.cid.as_ref().unwrap();
        if self.data.contains_key(cid) {
            let data = req.body.as_ref().unwrap();
            match DataClient::from_string(data.to_string()) {
                Ok(data) => {
                    self.data.insert(cid.to_string(), data);
                    return Response::Sucess("Client updated with sucess!".to_string());
                },
                Err(e) => {
                    return Response::Error(e.to_string());
                }
            }
        }
        return Response::Error("Client not found".to_string());
    }

    pub fn get(&mut self, req: &Request) -> Response {
        if req.cid == None{
            return Response::Error("Incomplet argumets".to_string());
        }

        let cid = req.cid.as_ref().unwrap();
        if self.data.contains_key(cid) {
            let data = self.data.get(cid).unwrap();

            match DataClient::to_string(data) {
                Ok(data) => {
                    return Response::Sucess(data);
                },    
                Err(e) => {
                    return Response::Error(e.to_string());
                }
            }
        }
        return Response::Error("Client not found".to_string());
    }

    pub fn delete(&mut self, req: &Request) -> Response {
        if req.cid == None{
            return Response::Error("Incomplet argumets".to_string());
        }

        let cid = req.cid.as_ref().unwrap();
        if self.data.contains_key(cid) {
            self.data.remove(cid);
            return Response::Sucess("Cliet deleted with sucess".to_string());
        }
        return Response::Error("Client not found".to_string())
    }
}

File: src/admin/controller.py
import sys
import plyvel

sys.path.append('.')

from others.models import DataClient
from others.response import Response, ResponseType
from others.request import Request, RequestType

class ClientController():
    def hashcode(data: str):
        PRIME = 18446744069414584321
        BASE = 5926535897

        acc = 1
        hashcode = 9471413089
        for char in bytes(data, 'utf-8'):
            acc = (acc * BASE) % PRIME
            hashcode = (hashcode + char * acc) % PRIME
        return hashcode

    def __init__(self, db: plyvel.DB):
        self.db = db
    
    def create(self, req: Request) -> Response:
        if req.req != RequestType.CreateClient:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.body == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        try:
            data_client = DataClient.from_string(req.body)  
        except Exception as e:
            return Response(ResponseType.Error, str(e))

        cid = ClientController.hashcode(req.body)
        while self.db.get(bytes(f'/{cid}', 'utf-8')) != None:
            cid += 1

        try:
            self.db.put(bytes(f'/{cid}', 'utf-8'), bytes(req.body, 'utf-8'))
            return Response(ResponseType.Sucess, f'Client created with sucess! Cid: {cid}')
        except Exception as e:
            return Response(ResponseType.Error, str(e))

    def update(self, req: Request) -> Response:
        if req.req != RequestType.UpdateClient:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.body == None or req.cid == None:
            return Response(ResponseType.Error, 'Incomplet argumets')

        try:
            data_client = DataClient.from_string(req.body)
        except Exception as e:
            return Response(ResponseType.Error, str(e))

        cid = req.cid
        if self.db.get(bytes(f'/{cid}', 'utf-8')):
            try:
                self.db.put(bytes(f'/{cid}', 'utf-8'), bytes(req.body, 'utf-8'))
                return Response(ResponseType.Sucess, 'Client updated with sucess')
            except Exception as e:
                return Response(ResponseType.Error, str(e))
        return Response(ResponseType.Error, 'Client not found')

    def get(self, req: Request) -> Response:
        if req.req != RequestType.GetClient:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None:
            return Response(ResponseType.Error, 'Incomplet argumets')
        
        cid = req.cid
        value = self.db.get(bytes(f'/{cid}', 'utf-8'))
        if value:
            value = str(value, 'utf-8')
            return Response(ResponseType.Sucess, value)
        return Response(ResponseType.Error, 'Client not found')

    def delete(self, req: Request) -> Response:
        if req.req != RequestType.DeleteClient:
            return Response(ResponseType.Error, 'Invalid Resquest')
        if req.cid == None:
            return Response(ResponseType.Error, 'Incomplet argumets')

        cid = req.cid                  
        try:
            self.db.delete(bytes(f'/{cid}', 'utf-8'))
            return Response(ResponseType.Sucess, 'Client deleted with sucess')
        except Exception as e:
            return Response(ResponseType.Error, 'Client not found')
